Terraform crash » Terraform v0.15.4 on linux_amd64 (docker:

<!--
Hi there,

Thank you for opening an issue. Please note that we try to keep the Terraform issue tracker reserved for bug reports and feature requests. For general usage questions, please see: https://www.terraform.io/community.html.

If your issue relates to a specific Terraform provider, please open it in the provider's own repository. The index of providers is at https://github.com/terraform-providers.

To fix problems, we need clear reproduction cases - we need to be able to see it happen locally. A reproduction case is ideally something a Terraform Core engineer can git-clone or copy-paste and run immediately, without inventing any details or context. 

* A short example can be directly copy-pasteable; longer examples should be in separate git repositories, especially if multiple files are needed
* Please include all needed context. For example, if you figured out that an expression can cause a crash, put the expression in a variable definition or a resource
* Set defaults on (or omit) any variables. The person reproducing it should not need to invent variable settings
* If multiple steps are required, such as running terraform twice, consider scripting it in a simple shell script. For example, see [this case](https://github.com/danieldreier/terraform-issue-reproductions/tree/master/25719). Providing a script can be easier than explaining what changes to make to the config between runs.
* Omit any unneeded complexity: remove variables, conditional statements, functions, modules, providers, and resources that are not needed to trigger the bug
* When possible, use the [null resource](https://www.terraform.io/docs/providers/null/resource.html) provider rather than a real provider in order to minimize external dependencies. We know this isn't always feasible. The Terraform Core team doesn't have deep domain knowledge in every provider, or access to every cloud platform for reproduction cases.

-->

### Terraform Version
<!---
Run `terraform version` to show the version, and paste the result between the ``` marks below.

If you are not running the latest version of Terraform, please try upgrading because your issue may have already been fixed.
-->

```
Terraform v0.15.4
on linux_amd64
+ provider registry.terraform.io/hashicorp/archive v2.2.0
+ provider registry.terraform.io/hashicorp/aws v3.42.0
+ provider registry.terraform.io/hashicorp/random v2.3.1
+ provider registry.terraform.io/hashicorp/template v2.2.0
```

Note: using container `hashicorp/terraform:light` or `hashicorp/terraform:0.15.4`

### Terraform Configuration Files
<!--
Paste the relevant parts of your Terraform configuration between the ``` marks below.

For Terraform configs larger than a few resources, or that involve multiple files, please make a GitHub repository that we can clone, rather than copy-pasting multiple files in here. For security, you can also encrypt the files using our GPG public key at https://www.hashicorp.com/security.
-->

```terraform
...
```

### Debug Output
<!--
Full debug output can be obtained by running Terraform with the environment variable `TF_LOG=trace`. Please create a GitHub Gist containing the debug output. Please do _not_ paste the debug output in the issue, since debug output is long.

Debug output may contain sensitive information. Please review it before posting publicly, and if you are concerned feel free to encrypt the files using the HashiCorp security public key.
-->

```
$ terraform -chdir=. plan -var "assume_role_arn=$TF_ASSUME_ROLE_ARN" -var "remote_state_token=$TF_TOKEN" -var-file=production.tfvars $OUTPUT
```

```
Acquiring state lock. This may take a few moments...
--
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0xbdf185]
 
goroutine 1516 [running]:
github.com/hashicorp/terraform/internal/terraform.(*NodeAbstractResourceInstance).planDestroy(0xc0011b56c0, 0x2cef4c0, 0xc001a847e0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
/home/circleci/project/project/internal/terraform/node_resource_abstract_instance.go:402 +0x185
github.com/hashicorp/terraform/internal/terraform.(*NodePlannableResourceInstanceOrphan).managedResourceExecute(0xc00295a5e0, 0x2cef4c0, 0xc001a847e0, 0x296a1b8, 0xc000066800, 0xc0021b1c50)
/home/circleci/project/project/internal/terraform/node_resource_plan_orphan.go:126 +0x538
github.com/hashicorp/terraform/internal/terraform.(*NodePlannableResourceInstanceOrphan).Execute(0xc00295a5e0, 0x2cef4c0, 0xc001a847e0, 0xc000100002, 0xc0021b1ce8, 0x40bb05, 0x24a95e0)
/home/circleci/project/project/internal/terraform/node_resource_plan_orphan.go:49 +0xb3
github.com/hashicorp/terraform/internal/terraform.(*ContextGraphWalker).Execute(0xc00272aa90, 0x2cef4c0, 0xc001a847e0, 0x7fbc71dde0a0, 0xc00295a5e0, 0x0, 0x0, 0x0)
/home/circleci/project/project/internal/terraform/graph_walk_context.go:129 +0xbf
github.com/hashicorp/terraform/internal/terraform.(*Graph).walk.func1(0x27fd220, 0xc00295a5e0, 0x0, 0x0, 0x0)
/home/circleci/project/project/internal/terraform/graph.go:59 +0xbd3
github.com/hashicorp/terraform/internal/dag.(*Walker).walkVertex(0xc00038bb00, 0x27fd220, 0xc00295a5e0, 0xc000965680)
/home/circleci/project/project/internal/dag/walk.go:381 +0x288
created by github.com/hashicorp/terraform/internal/dag.(*Walker).Update
/home/circleci/project/project/internal/dag/walk.go:304 +0x1246
 
 
 
 
!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!
 
Terraform crashed! This is always indicative of a bug within Terraform.
A crash log has been placed at "crash.020280254.log" relative to your current
working directory. It would be immensely helpful if you could please
report the crash with Terraform[1] so that we can fix this.
 
When reporting bugs, please include your terraform version. That
information is available on the first line of crash.log. You can also
get it by running 'terraform --version' on the command line.
 
SECURITY WARNING: the "crash.020280254.log" file that was created may contain
sensitive information that must be redacted before it is safe to share
on the issue tracker.
 
[1]: https://github.com/hashicorp/terraform/issues
 
!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!
```

### Crash Output
<!--
If the console output indicates that Terraform crashed, please share a link to a GitHub Gist containing the output of the `crash.log` file.
-->

Please provide a private channel for `crash.*.log` file.

### Expected Behavior
<!--
What should have happened?
-->

The plan will plan.

### Actual Behavior
<!--
What actually happened?
-->

Crashed and told to contact.

Note: works on v0.14.11 and v0.15.3.

### Steps to Reproduce
<!--
Please list the full steps required to reproduce the issue, for example:
1. `terraform init`
2. `terraform apply`
-->

### Additional Context
<!--
Are there anything atypical about your situation that we should know? For example: is Terraform running in a wrapper script or in a CI system? Are you passing any unusual command line options or environment variables to opt-in to non-default behavior?
-->

Docker container `hashicorp/terraform:light` in CI/CD.

### References
<!--
Are there any other GitHub issues (open or closed) or Pull Requests that should be linked here? For example:

- #6017

-->

--- a/internal/terraform/node_resource_abstract_instance.go
+++ b/internal/terraform/node_resource_abstract_instance.go
@@ -399,7 +399,6 @@ func (n *NodeAbstractResourceInstance) planDestroy(ctx EvalContext, currentState
 				Before: cty.NullVal(cty.DynamicPseudoType),
 				After:  cty.NullVal(cty.DynamicPseudoType),
 			},
-			Private:      currentState.Private,
 			ProviderAddr: n.ResolvedProvider,
 		}
 		return noop, nil

--- a/internal/terraform/transform_orphan_resource.go
+++ b/internal/terraform/transform_orphan_resource.go
@@ -79,7 +79,12 @@ func (t *OrphanResourceInstanceTransformer) transform(g *Graph, ms *states.Modul
 			}
 		}
 
-		for key := range rs.Instances {
+		for key, inst := range rs.Instances {
+			// deposed instances will be taken care of separately
+			if inst.Current == nil {
+				continue
+			}
+
 			addr := rs.Addr.Instance(key)
 			abstract := NewNodeAbstractResourceInstance(addr)
 			var node dag.Vertex = abstract

--- a/internal/terraform/transform_orphan_resource_test.go
+++ b/internal/terraform/transform_orphan_resource_test.go
@@ -50,6 +50,26 @@ func TestOrphanResourceInstanceTransformer(t *testing.T) {
 				Module:   addrs.RootModule,
 			},
 		)
+
+		// A deposed orphan should not be handled by this transformer
+		s.SetResourceInstanceDeposed(
+			addrs.Resource{
+				Mode: addrs.ManagedResourceMode,
+				Type: "test_instance",
+				Name: "deposed",
+			}.Instance(addrs.NoKey).Absolute(addrs.RootModuleInstance),
+			states.NewDeposedKey(),
+			&states.ResourceInstanceObjectSrc{
+				AttrsFlat: map[string]string{
+					"id": "foo",
+				},
+				Status: states.ObjectReady,
+			},
+			addrs.AbsProviderConfig{
+				Provider: addrs.NewDefaultProvider("test"),
+				Module:   addrs.RootModule,
+			},
+		)
 	})
 
 	g := Graph{Path: addrs.RootModuleInstance}

